limiter.go

```diff
@@ -144,6 +144,15 @@ func (dl *dialLimiter) AddDialJob(dj *dialJob) {
 	go dl.executeDial(dj)
 }
 
+func (dl *dialLimiter) clearAllPeerDials(p peer.ID) {
+	dl.rllock.Lock()
+	defer dl.rllock.Unlock()
+	delete(dl.waitingOnPeerLimit, p)
+	// NB: the waitingOnFd list doesnt need to be cleaned out here, we will
+	// remove them as we encounter them because they are 'cancelled' at this
+	// point
```

Comment:
Same as the `waitingOnPeerLimit` list. The problem is that it could be a *while* before that happens if we're really backed up behind a bunch of slow dials. It won't be *quite* as bad because it's limited by `dials*perPeerLimit` but it will still be large and can still grow in the same way.

Reply:
@Stebalien Yeah, the tradeoff we make is the O(n) operation under a held lock after completion  vs this potential extra memory buffering.

Reply:
That could be solved with a doubly linked list (wait queues like this are really one of the only uses for them).

Reply:
Ah, great point. We never really need to iterate over it, so that makes perfect sense.